src/lib/libast/man/aso.3
.TH ASO 3
.SH NAME
\fBASO\fP \- Atomic Scalar Operations
.SH SYNOPSIS
.de Tp
.fl
.ne 2
.TP
..
.de Ss
.fl
.ne 2
.SS "\\$1"
..
.de Cs
.nf
.ft CR
..
.de Ce
.ft 1
.fi
..
.ta 1.0i 2.0i 3.0i 4.0i 5.0i
.Cs
#include <aso.h>
.Ce
.SS "TYPES"
.Cs
typedef int (*Asoerror_f)(int, const char*);
typedef void* (*Asoinit_f)(void*, const char*);
typedef ssize_t (*Asolock_f)(void*, ssize_t, void volatile*);

typedef struct Asodisc_s
{
	uint32_t	version;
	unsigned int	hung;
	Asoerror_f	errorf;
} Asodisc_t;

typedef struct Asometh_s
{
	const char*	name;
	int		type;
	Asoinit_f	initf;
	Asolock_f	lockf;
	const char*	details;
} Asometh_t;
.Ce
.SS "OPERATIONS"
.Cs
uint8_t		asocas8(uint8_t volatile*, int, int);
uint8_t		asoget8(uint8_t volatile*);
uint8_t		asoinc8(uint8_t volatile*);
uint8_t		asodec8(uint8_t volatile*);

uint16_t	asocas16(uint16_t volatile*, uint16_t, uint16_t);
uint16_t	asoget16(uint16_t volatile*);
uint16_t	asoinc16(uint16_t volatile*);
uint16_t	asodec16(uint16_t volatile*);

uint32_t	asocas32(uint32_t volatile*, uint32_t, uint32_t);
uint32_t	asoget32(uint32_t volatile*);
uint32_t	asoinc32(uint32_t volatile*);
uint32_t	asodec32(uint32_t volatile*);

uint64_t	asocas64(uint64_t volatile*, uint64_t, uint64_t);
uint64_t	asoget64(uint64_t volatile*);
uint64_t	asoinc64(uint64_t volatile*);
uint64_t	asodec64(uint64_t volatile*);

unsigned char	asocaschar(unsigned char volatile*, int, int);
unsigned char	asogetchar(unsigned char volatile*);
unsigned char	asoincchar(unsigned char volatile*);
unsigned char	asodecchar(unsigned char volatile*);

unsigned short	asocasshort(unsigned short volatile*, unsigned short, unsigned short);
unsigned short	asogetshort(unsigned short volatile*);
unsigned short	asoincshort(unsigned short volatile*);
unsigned short	asodecshort(unsigned short volatile*);

unsigned int	asocasint(unsigned int volatile*, unsigned int, unsigned int);
unsigned int	asogetint(unsigned int volatile*);
unsigned int	asoincint(unsigned int volatile*);
unsigned int	asodecint(unsigned int volatile*);

unsigned long	asocaslong(unsigned long volatile*, unsigned long, unsigned long);
unsigned long	asogetlong(unsigned long volatile*);
unsigned long	asoinclong(unsigned long volatile*);
unsigned long	asodeclong(unsigned long volatile*);

size_t		asocassize(size_t volatile*, size_t, size_t);
size_t		asogetsize(size_t volatile*);
size_t		asoincsize(size_t volatile*);
size_t		asodecsize(size_t volatile*);

void*		asocasptr(void volatile*, void*, void*);
void*		asogetptr(void volatile*);

void		ASODISC(Asodisc_t*, Asoerror_f);
Asometh_t*	asometh(int, void*);
int		asoinit(const char*, Asometh_t*, Asodisc_t*);
int		asolock(unsigned int volatile*, unsigned int, int);
int		asoloop(uintmax_t);
int		asorelax(uint32_t);
.Ce
.SH DESCRIPTION
\fIASO\fP provides functions to perform atomic scalar operations.
The functions on the type \f3uint32_t\fP will be fully described below.
Other functions work similarly on their respective types.
Some of the functions may be macros that call other functions.
64 bit operations are provided if the compiler supports 64 bit integers
and/or pointers.
.SS "TYPES"
\f3uint8_t, uint16_t, uint32_t, uint64_t\fP

These are \fIunsigned integer\fP types of different sizes in bits.
For example, \f3uint32_t\fP represents the type of unsigned integer
with 32 bits or 4 bytes.
.SS "OPERATIONS"
.SS "  uint32_t asoget32(uint32_t* from);"
This function returns the value \f3*from\fP.
.SS "  uint32_t asoinc32(uint32_t* dest);"
.SS "  uint32_t asodec32(uint32_t* dest);"
These functions increment \f3*dest\fP by 1 and decrement \f3*dest\fP by 1
in an atomic step.
The return value is the old value in \f3*dest\fP.

Consider an example where two concurrent threads/processes call \f3asoinc32()\fP
on the same \f3dest\fP with values, say \fIv1\fP and \fIv2\fP.
The eventual value in \f3dest\fP
will be as if \f3*dest += 2\fP was performed in a single-threaded execution.

That should be contrasted with a situation where, instead of \f3asoinc32()\fP
or \f3asodec32()\fP,
only normal increment (++) or decrement (--) were used.
Then, the end result could be either \f3*dest += 1\fP or \f3*dest += 2\fP,
depending on states of the hardware cache and process scheduling.
.SS "  uint32_t asocas32(uint32_t* dest, uint32_t tstval, uint32_t newval);"
This function provides the atomic \fIcompare-and-swap\fP operation.
If the current content of \f3dest\fP is equal to \f3tstval\fP then it will be
set to \f3newval\fP.
If multiple threads/processes are performing the same operations only one
will succeed with a return value of \f3tstval\fP.
The return value is the old value in \f3*dest\fP.
.SS "  void asorelax(uint32_t nsec)"
This function causes the calling process or thread to briefly pause
for \f3nsec\fP nanoseconds.
It is useful to implement tight loops that occasionally yield control.
.SS "  int asolock(unsigned int* lock, unsigned int key, int type)"
This function uses \f3key\fP, a non-zero unsigned integer, to lock or
unlock the \f3lock\fP.
It returns \f30\fP on success and \f3-1\fP on failure.
The argument \f3type\fP can take one of the following values:
.Tp
\f3ASO_UNLOCK\fP:
This unlocks the lock if it was locked with \f3key\fP. It is an error to try
unlocking a lock of a different key.
.Tp
\f3ASO_TRYLOCK\fP:
This makes a single attempt to use the given \f3key\fP to acquire a lock.
An error will result if the lock is already locked with a different key.
.Tp
\f3ASO_LOCK\fP:
This is a regular locking call. If the lock is locked with a different key,
this call will wait until the lock is open, then lock it with
the given \f3key\fP.
.Tp
\f3ASO_SPINLOCK\fP:
Regardless of what key is currently locking the lock,
this call will always wait until the lock is open, then lock it with
the given \f3key\fP.
Note that, if the lock is already locked with \f3key\fP, this call can result
in a deadlock unless that lock can be opened by some other mechanism, e.g.,
by a different process or thread.
.SS "  int asoloop(uintmax_t iteration);"
This function is used to implement spin locks that periodically relinquish
the processor:
.Cs
uintmax_t iteration;
iteration = 0;
for (;;) {
    /* test resource with an aso*() call */
    if (asoloop(++iteration))
        /* an error occurred */;
}
.Ce
The value of \f3iteration\fP should be \f31\fP (\fInot\fP \f30\fP) for the
first loop iteration.
\f30\fP is returned on success, \f3-1\fP on failure.
If \f3iteration mod 4\fP is \f30\fP then \f3asorelax(1)\fP is called to
temporarily relinquish
the processor.
If \f3Asodisc_t.hung != 0\fP and \f3Asodisc_t.errorf != 0\fP and
\f3iteration mod (2**Asodisc_t.hung-1)\fP is \f30\fP,
then \f3Asodisc_t.errorf\fP is called with type \f3ASO_HUNG\fP
and \f3-1\fP is returned.
.SS "DISCIPLINE"
The Asodisc_t discipline structure allows the caller to modify default behavior.
The \fIASO\fP discipline is global for all threads and forked children of the
current process.
The discipline is set and modified by the \f3asoinit()\fP function, described
below.
The structure members are:
.Tp
\f3uint32_t version;\fP
This must be set to \f3ASO_VERSION\fP by the caller and is used by the
implementation to detect
release differences between the caller and the implementation.
The version is integer of the form \fIYYYYMMDD\fP where \fIYYYY\fP is the
release year, \fIMM\fP
is the release month, and \fIDD\fP is the release day of month.
This allows the implementation to be forwards and backwards binary compatible
with all releases.
.Tp
\f3unsigned int hung;\fP
An error occurs if \f3asoloop\fP() is called \f32**Asometh_t.hung\fP times
without gaining access to the loop resource.
The default value \f30\fP disables the test.
.Tp
\f3Asoerror_f errorf;\fP
\f3int (*errorf)(int type, const char* mesg);\fP
If \f3errorf\fP != \f30\fP then it is called for each \fIASO\fP fatal
library condition.
\f3type\fP may be one of: \f3ASO_METHOD\fP - a method error;
\f3ASO_HUNG\fP - \f3asoloop\fP() was called
\f32**Asometh_t.hung\fP times with no access to the loop resource.
\f3mesg\fP is a 0-terminated message description.
.SS "  void ASODISC(Asodisc_t* disc, Asoerror_f errorf);"
This function-like-macro initializes
\f3disc->version = ASO_VERSION\fP, \f3disc->errorf = errorf\fP,
and the remaining \f3disc\fP members to \f30\fP.
.SS "METHODS"
Several atomic locking methods are implemented for atomic operations
not supported by \fIintrinsic\fP functions or assembly instructions.
Methods are controlled by the \f3asometh()\fP and \f3asoinit()\fP
functions, described below.
The \fIASO\fP method is global for all threads and forked children of
the current process.
A given method may have multiple types.
The methods types are:
.Tp
\f3ASO_INTRINSIC\fP:
Some hardware platforms provide machine instructions to implement
these operations directly.
In that case, if a local compiler permits, calls to these
\fIintrinsic\fP functions
may be translated directly into their corresponding machine instructions.
When necessary the implementation can use only the
intrinsic \fIcompare-and-swap\fP
function on the largest integer type to emulate all other \fIASO\fP operations.
The \f3ASO_INTRINSIC\fP method type is the default when supported
by the compiler.
It may be used for single-process single-thread, multi-thread, and
multi-process applications.
When supported by the hardware / compiler, the library provides the
"\fBintrinsic\fP" method with type
\f3ASO_INTRINSIC|ASO_PROCESS|ASO_THREAD|ASO_SIGNAL\fP.
.Tp
\f3ASO_SIGNAL\fP:
This method type is suitable only for single-process
single-thread applications.
It can be used to provide locking between asynchronous \fBsignal\fP(2) handlers
and the main program.
The library provides the "\fBsignal\fP" method with type \f3ASO_SIGNAL\fP.
This is the default method type when \f3ASO_INTRINSIC\fP is not supported.
.Tp
\f3ASO_THREAD\fP:
This method type is suitable for single-process single-thread, and
multi-thread applications.
It typically requires thread library support, and since the default
\f3aso\fP code in \fBlibast\fP is not linked with a thread library,
no \f3ASO_THREAD\fP method is provided by default.
The "\fBspin\fP" (using \fBpthread_spin_lock\fP(3)) and
"\fBmutex"\fP (using \fBpthread_mutex_lock\fP(3)) methods with
type \f3ASO_THREAD|ASO_SIGNAL\fP
can be made available by intercepting \f3asometh()\fP
to prepend them ahead of the default methods.
.Tp
\f3ASO_PROCESS\fP:
This method type is suitable for single-process single-thread,
and multi-process applications.
Some \f3ASO_PROCESS\fP methods may also be suitable for multi-thread

NEWS
This documents significant changes in the dev branch of ksh 93u+m.
For full details, see the git log at: https://github.com/ksh93/ksh
Uppercase BUG_* IDs are shell bug IDs as used by the Modernish shell library.

2026-09-18:

- Fixed: a command substitution will now correctly capture the output of
  background processes launched within it but not explicitly waited for.

2026-09-17:

- Fixed multiple memory leaks occurring upon subshell exit when assigning a
  value to a variable with a shell discipline function within a subshell.

- Fixed a crash occurring when assigning, within a subshell, an invalid
  value to a variable of a type defined by the 'enum' command.

2026-09-16:

- Fixed: when the special IFS (field splitting characters) variable was
  assigned to within a ${ shared-state command substitution }, the state of IFS
  failed to persist beyond it, and a use after free (potential crash) occurred.

- Fixed bounds checking when assigning to fixed-size arrays (declared like,
  e.g., 'typeset -a array[4]'). Assigning to an out-of-range subscript now
  produces the expected error instead of potentially crashing.

2026-09-15:

- Fixed a memory leak when assigning, within a subshell, a value to a
  special variable (one of IFS, FPATH, CDPATH, SHELL, ENV, VISUAL, EDITOR,
  HISTFILE, OPTIND, SECONDS, RANDOM,.sh.match, .sh.version, LINENO, _,
  LC_CTYPE, LC_NUMERIC, LC_COLLATE, LC_MESSAGES, LC_ALL, and LANG).

2026-09-13:

- Fixed a memory leak that occurred when converting an existing scalar
  variable to an associative array using 'typeset -A'.

2026-09-12:

- Implemented advisory locking for $HISTFILE (~/.sh_history by default) so
  that multiple ksh instances can write to the history file without risking
  data loss when one of them decides to trim the file. Note that ksh
  versions before 2026-09-12 will ignore the lock, so the risk continues to
  exist if an older ksh binary shares write access to the file.

- History file trimming no longer requires write access to the directory in
  which the history file resides and will no longer change the directory's
  last-modified timestamp. However, it now requires a usable /tmp or $TMPDIR.

- I/O errors while writing to the history file are no longer silently ignored.

- Fixed a possible crash when attempting to set an integer variable to a
  multibyte character.

2026-09-10:

- Fixed a rare crash that could occur if a signal (such as interrupt, Ctrl+C)
  occurred while cleaning up the state of a virtual subshell upon its exit.

2026-09-08:

- Fixed: PWD and ENV were not properly reinitialised when executing a shell
  script without a #! path.

- Fixed a possible crash after unsetting RANDOM. Bug introduced on 2022-07-21.

2026-09-04:

- Fixed a longstanding bug in which output data loss or corruption could
  intermittently occur if, within a command substitution, standard error is
  redirected to standard output and, within that redirection, standard
  output is then connected to a pipe.

2026-09-02:

- Fixed a bug, introduced on 2025-05-30, that made 'print -s' ineffective
  at writing to the history file. This functionality is now restored.

2026-08-31:

- Fixed a crash in the expr(1) path-bound built-in command that could occur
  when using the non-standard 'match' or 'index' keyword.

2026-08-30:

- Fixed a bug that caused non-importable environment variables (i.e., those
  with names that are not valid shell variable names) to disappear from the
  environment of child processes after invoking more than two external
  commands, potentially crashing the shell. Bug introduced on 2025-05-30.

2026-08-29:

- If an error (such as a syntax error) occurs in a trap action, the error
  message will now have the word "(trap)" in the call stack for clarity.

2026-08-26:

- Following up to 2026-08-16, fixed a remaining problem in the 'read'
  built-in with combined -d and -n options when reading from a pipe; the
  custom delimiter is now correctly used instead of newline.

2026-08-16:

- Fixed a long-standing bug in the 'read' built-in: the -d option was
  ignored when combined with -n. The two options can now be used together.

2026-08-14:

- In the --man/--help output for built-in commands, text marked as italic
  is no longer rendered as bold + underlined, but only as underlined.

2026-08-13:

- Fixed corruption of $_, crashing the shell upon 'typeset -p _', after an
  error occurs trying to redefine a type defined with 'typeset -T'.

2026-08-06:

- Fixed ksh crashing upon parsing a very long list of simple commands
  separated by semicolons or, in 'eval' or dot script contexts, by newlines.

- [v1.1] The -n/--noexec option can now effectively be set in subshells of
  interactive shells, e.g., within a parentheses block.

- [v1.1] The 'eval' command has a new -n/--noexec option that causes it to
  subject the command(s) given to static analysis in the manner of the
  -n/--noexec shell option, emitting warnings for deprecated usages.

2026-08-04:

- Fixed a potential crash when suspending (^Z), on the interactive shell,
  a multi-element pipeline with the last element running in the current
  environment.

- For non-interactive shells, the COLUMNS and LINES variables now initially

src/lib/libast/aso/aso.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1985-2012 AT&T Intellectual Property          *
*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 2.0                  *
*                                                                      *
*                A copy of the License is available at                 *
*      https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html      *
*         (with md5 checksum 84283fa8859daf213bdda5a9f8d1be1d)         *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                   Phong Vo <kpv@research.att.com>                    *
*                  Martijn Dekker <martijn@inlv.org>                   *
*            Johnothan King <johnothanking@protonmail.com>             *
*                     Cy Schubert <cy@FreeBSD.org>                     *
*       hyousatsu <118750527+hyousatsu@users.noreply.github.com>       *
*                                                                      *
***********************************************************************/

#include "asohdr.h"
#include "FEATURE/aso"

/*
 * AST atomic scalar operations
 * AT&T Research
 *
 * cas { 8 16 32 [64] } subset snarfed from the work by
 * Adam Edgar and Kiem-Phong Vo 2010-10-10
 *
 * lock methods and emulations by
 * Glenn Fowler 2011-11-11
 *
 * hopefully stable by 2012-12-12
 */

#if defined(_aso_casptr) && (defined(_aso_cas32) || defined(_aso_cas64))
#define ASO_METHOD		(&_aso_meth_intrinsic)
#define ASO_LOCKF		0
#else
#define ASO_METHOD		(&_aso_meth_signal)
#define ASO_LOCKF		_aso_lock_signal
#endif

typedef union
{
	uint8_t			c[2];
	uint16_t		i;
} U16_8_t;

typedef union
{
	uint8_t			c[4];
	uint32_t		i;
} U32_8_t;

typedef union
{
	uint16_t		c[2];
	uint32_t		i;
} U32_16_t;

#ifdef _ast_int8_t

typedef union
{
	uint8_t			c[8];
	uint64_t		i;
} U64_8_t;

typedef union
{
	uint16_t		c[4];
	uint64_t		i;
} U64_16_t;

typedef union
{
	uint32_t		c[2];
	uint64_t		i;
} U64_32_t;

#endif

typedef struct State_s
{
	Asometh_t*		meth;
	Asolock_f		lockf;
	Asoerror_f		errorf;
	uintmax_t		hung;
	unsigned int		hung2;
	void*			data;
	pid_t			pid;
} State_t;

static unsigned int		_aso_data_signal;

static ssize_t
_aso_lock_signal(void* data, ssize_t k, void volatile* p)
{
	NOT_USED(data);
	NOT_USED(p);
	if (k >= 0)
	{
		_aso_data_signal--;
		return 0;
	}
	while (_aso_data_signal++)
		_aso_data_signal--;
	return 1;
}

static Asometh_t	_aso_meth_signal =    { "signal",    ASO_SIGNAL,    0, _aso_lock_signal };
extern Asometh_t	_aso_meth_semaphore;
extern Asometh_t	_aso_meth_fcntl;
static Asometh_t	_aso_meth_intrinsic = { "intrinsic", ASO_INTRINSIC|ASO_PROCESS|ASO_THREAD|ASO_SIGNAL, 0, 0 };

static Asometh_t*	method[] =
{
	&_aso_meth_signal,
#if defined(_ast_int8_t) && defined(_aso_cas64) || !defined(_ast_int8_t) && defined(_aso_cas32)
	&_aso_meth_intrinsic,
#endif
#if _aso_semaphore
	&_aso_meth_semaphore,
#endif
#if _aso_fcntl
	&_aso_meth_fcntl,
#endif
};

static State_t			state =
{
	ASO_METHOD, ASO_LOCKF
};

static COLD int
asoerror(int type, const char* format, const char* a, const char* b, unsigned int n)
{
	char	buf[128];

	if (b)
		sfsprintf(buf, sizeof(buf), format, a, b, n);
	else if (a)
		sfsprintf(buf, sizeof(buf), format, a, n);
	else
		sfsprintf(buf, sizeof(buf), format, n);
	return state.errorf(type, buf);
}

/*
 * if type!=0 return lock method for type with name details
 * else if name!=0 return lock method matching <name>[,<details>]
 * else return the current lock method
 * 0 returned on error
 *
 * the user visible asometh() calls this function
 * it allows, e.g., for -ltaso to provide an asometh() intercept
 * that prepends its ASO_THREAD methods ahead of the _asometh() methods
 */

Asometh_t*
_asometh(int type, void* data)
{
	size_t		n;
	size_t		i;
	char*		e;
	Asometh_t*	meth;
	char*		name;

	if (type == ASO_NEXT)
	{
		if (!(meth = (Asometh_t*)data))
			return method[0];
		for (i = 0; i < elementsof(method) - 1; i++)
			if (meth == method[i])
				return method[i+1];
		return NULL;
	}
	if (type)
	{
		for (i = 0; i < elementsof(method); i++)
			if (method[i]->type & type)
			{
				method[i]->details = (char*)data;
				return method[i];
			}
		return NULL;
	}
	if (!(name = (char*)data))
		return state.meth;
	n = (e = strchr(name, ',')) ? (size_t)(e - name) : strlen(name);
	for (i = 0; i < elementsof(method); i++)
		if (strncmp(name, method[i]->name, n) == 0)
		{
			if (e)
				method[i]->details = e + 1;
			return method[i];
		}
	return NULL;
}

/*
 * clean up lock method on exit
 */

static void
asoexit(void)
{
	if (state.meth && state.meth->initf && state.data && state.pid == getpid())
	{
		state.lockf = ASO_METHOD->lockf;
		state.meth->initf(state.data, 0);
		state.data = 0;
	}
}

/*
 * initialize lock method
 */

int
asoinit(const char* details, Asometh_t* meth, Asodisc_t* disc)
{
	void*		data;

	if (disc)
	{
		state.errorf = disc->errorf;
		state.hung2 = disc->hung;
		state.hung = 1;
		state.hung <<= state.hung2;
		state.hung--;
	}
	if (!meth)
		return state.pid != 0;
	if (!meth->lockf && !(meth->type & ASO_INTRINSIC))
	{
		if (state.errorf)
			asoerror(ASO_EMETHOD, "%s method has no lock function", meth->name, 0, 0);
		return -1;
	}
	state.lockf = ASO_METHOD->lockf;
	if (state.meth && state.meth->initf && state.data)
	{
		state.meth->initf(state.data, 0);
		state.data = 0;
	}
	if (!meth->initf)
		data = 0;
	else if (!(data = meth->initf(0, details ? details : meth->details)))
	{
		state.meth = ASO_METHOD;
		if (state.errorf)
			asoerror(ASO_EMETHOD, "%s method initialization failed -- reverting to the %s method", meth->name, state.meth->name, 0);
		return -1;
	}
	state.meth = meth;
	state.data = data;
	state.lockf = meth->lockf;
	if (!state.pid)
	{
		state.pid = getpid();
		atexit(asoexit);
	}
	return 0;
}

/*
 * loop check for hung spin locks
 * and periodic relinquishing of the processor
 */

int
asoloop(uintmax_t rep)
{
	if (state.hung && !(rep & state.hung) && state.errorf)
		return asoerror(ASO_EHUNG, "spin lock possibly hung after 2^%u attempts", 0, 0, state.hung2);
	return (rep & ASO_RELAX) ? 0 : asorelax(1);
}

/*
 * error checking state.lockf() call
 */

static ssize_t
lock(void* data, ssize_t k, void volatile* p)
{
	ssize_t		r;

	if ((r = state.lockf(data, k, p)) < 0 && state.errorf)
		asoerror(ASO_EMETHOD, "%s method lock failed", state.meth->name, 0, 0);
	return r;
}

/*
 * sync and return "current" value
 */

uint8_t
asoget8(uint8_t volatile* p)
{
	uint8_t		o;

	do
	{
		o = *p;
	} while (asocas8(p, o, o) != o);
	return o;
}

uint16_t
asoget16(uint16_t volatile* p)
{
	uint16_t	o;

	do
	{
		o = *p;
	} while (asocas16(p, o, o) != o);
	return o;
}

uint32_t
asoget32(uint32_t volatile* p)
{
	uint32_t	o;

	do
	{
		o = *p;
	} while (asocas32(p, o, o) != o);
	return o;
}

#ifdef _ast_int8_t

uint64_t
asoget64(uint64_t volatile* p)
{
	uint64_t	o;

	do
	{
		o = *p;
	} while (asocas64(p, o, o) != o);
	return o;
}

#endif

void*
asogetptr(void volatile* p)
{
	void*	o;

	do
	{
		o = *(void* volatile*)p;
	} while (asocasptr(p, o, o) != o);
	return o;
}

/*
 * increment and return old value
 */

uint8_t
asoinc8(uint8_t volatile* p)
{
	ssize_t		k;
	uint8_t		o;

#if defined(_aso_inc8)
	if (!state.lockf)
		return _aso_inc8(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas8(p, o, o + 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)++;
	lock(state.data, k, p);
	return o;
}

uint16_t
asoinc16(uint16_t volatile* p)
{
	ssize_t		k;
	uint16_t	o;

#if defined(_aso_inc16)
	if (!state.lockf)
		return _aso_inc16(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas16(p, o, o + 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)++;
	lock(state.data, k, p);
	return o;
}

uint32_t
asoinc32(uint32_t volatile* p)
{
	ssize_t		k;
	uint32_t	o;

#if defined(_aso_inc32)
	if (!state.lockf)
		return _aso_inc32(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas32(p, o, o + 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)++;
	lock(state.data, k, p);
	return o;
}

#ifdef _ast_int8_t

uint64_t
asoinc64(uint64_t volatile* p)
{
	ssize_t		k;
	uint64_t	o;

#if defined(_aso_inc64)
	if (!state.lockf)
		return _aso_inc64(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas64(p, o, o + 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)++;
	lock(state.data, k, p);
	return o;
}

#endif

/*
 * decrement and return old value
 */

uint8_t
asodec8(uint8_t volatile* p)
{
	ssize_t		k;
	uint8_t		o;

#if defined(_aso_dec8)
	if (!state.lockf)
		return _aso_dec8(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas8(p, o, o - 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)--;
	lock(state.data, k, p);
	return o;
}

uint16_t
asodec16(uint16_t volatile* p)
{
	ssize_t		k;
	uint16_t	o;

#if defined(_aso_dec16)
	if (!state.lockf)
		return _aso_dec16(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas16(p, o, o - 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)--;
	lock(state.data, k, p);
	return o;
}

uint32_t
asodec32(uint32_t volatile* p)
{
	ssize_t		k;
	uint32_t	o;

#if defined(_aso_dec32)
	if (!state.lockf)
		return _aso_dec32(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas32(p, o, o - 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)--;
	lock(state.data, k, p);
	return o;
}

#ifdef _ast_int8_t

uint64_t
asodec64(uint64_t volatile* p)
{
	ssize_t		k;
	uint64_t	o;

#if defined(_aso_dec64)
	if (!state.lockf)
		return _aso_dec64(p);
#else
	if (!state.lockf)
	{
		do
		{
			o = *p;
		} while (asocas64(p, o, o - 1) != o);
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	o = (*p)--;
	lock(state.data, k, p);
	return o;
}

#endif

/*
 * { 8 16 32 [64] } compare with old, swap with new if same, and return old value
 */

uint8_t
asocas8(uint8_t volatile* p, uint8_t o, uint8_t n)
{
	ssize_t		k;

#if defined(_aso_cas8)
	if (!state.lockf)
		return _aso_cas8(p, o, n);
#elif defined(_aso_cas16)
	if (!state.lockf)
	{
		U16_8_t		u;
		U16_8_t		v;
		U16_8_t*	a;
		ssize_t		s;
		ssize_t		i;

		s = (ssize_t)(integralof(p) & (sizeof(u.i) - 1));
		a = (U16_8_t*)((char*)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
		for (;;)
		{
			u.i = a->i;
			u.c[s] = o;
			v.i = u.i;
			v.c[s] = n;
			if (_aso_cas16(&a->i, u.i, v.i) == u.i)
				break;
			for (i = 0;; i++)
				if (i >= elementsof(u.c))
					return a->c[s];
				else if (i != s && u.c[i] != a->c[i])
					break;
		}
		return o;
	}
#elif defined(_aso_cas32)
	if (!state.lockf)
	{
		U32_8_t		u;
		U32_8_t		v;
		U32_8_t*	a;
		ssize_t		s;
		ssize_t		i;

		s = (ssize_t)(integralof(p) & (sizeof(u.i) - 1));
		a = (U32_8_t*)((char*)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
		for (;;)
		{
			u.i = a->i;
			u.c[s] = o;
			v.i = u.i;
			v.c[s] = n;
			if (_aso_cas32(&a->i, u.i, v.i) == u.i)
				break;
			for (i = 0;; i++)
				if (i >= elementsof(u.c))
					return a->c[s];
				else if (i != s && u.c[i] != a->c[i])
					break;
		}
		return o;
	}
#elif defined(_aso_cas64)
	if (!state.lockf)
	{
		U64_8_t		u;
		U64_8_t		v;
		U64_8_t*	a;
		ssize_t		s;
		ssize_t		i;

		s = (ssize_t)(integralof(p) & (sizeof(u.i) - 1));
		a = (U64_8_t*)((char*)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
		for (;;)
		{
			u.i = a->i;
			u.c[s] = o;
			v.i = u.i;
			v.c[s] = n;
			if (_aso_cas64(&a->i, u.i, v.i) == u.i)
				break;
			for (i = 0;; i++)
				if (i >= elementsof(u.c))
					return a->c[s];
				else if (i != s && u.c[i] != a->c[i])
					break;
		}
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	if (*p == o)
		*p = n;
	else
		o = *p;
	lock(state.data, k, p);
	return o;
}

uint16_t
asocas16(uint16_t volatile* p, uint16_t o, uint16_t n)
{
	ssize_t		k;

#if defined(_aso_cas16)
	if (!state.lockf)
		return _aso_cas16(p, o, n);
#elif defined(_aso_cas32)
	if (!state.lockf)
	{
		U32_16_t	u;
		U32_16_t	v;
		U32_16_t*	a;
		ssize_t		s;
		ssize_t		i;

		s = (ssize_t)(integralof(p) & (sizeof(u.i) - 1)) / 2;
		a = (U32_16_t*)((char*)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
		for (;;)
		{
			u.i = a->i;
			u.c[s] = o;
			v.i = u.i;
			v.c[s] = n;
			if (_aso_cas32(&a->i, u.i, v.i) == u.i)
				break;
			for (i = 0;; i++)
				if (i >= elementsof(u.c))
					return a->c[s];
				else if (i != s && u.c[i] != a->c[i])
					break;
		}
		return o;
	}
#elif defined(_aso_cas64)
	if (!state.lockf)
	{
		U64_16_t	u;
		U64_16_t	v;
		U64_16_t*	a;
		ssize_t		s;
		ssize_t		i;

		s = (ssize_t)(integralof(p) & (sizeof(u.i) - 1)) / 2;
		a = (U64_16_t*)((char*)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
		for (;;)
		{
			u.i = a->i;
			u.c[s] = o;
			v.i = u.i;
			v.c[s] = n;
			if (_aso_cas64(&a->i, u.i, v.i) == u.i)
				break;
			for (i = 0;; i++)
				if (i >= elementsof(u.c))
					return a->c[s];
				else if (i != s && u.c[i] != a->c[i])
					break;
		}
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	if (*p == o)
		*p = n;
	else
		o = *p;
	lock(state.data, k, p);
	return o;
}

uint32_t
asocas32(uint32_t volatile* p, uint32_t o, uint32_t n)
{
	ssize_t		k;

#if defined(_aso_cas32)
	if (!state.lockf)
		return _aso_cas32(p, o, n);
#elif defined(_aso_cas64)
	if (!state.lockf)
	{
		U64_32_t	u;
		U64_32_t	v;
		U64_32_t*	a;
		ssize_t		s;
		ssize_t		i;

		s = (ssize_t)(integralof(p) & (sizeof(u.i) - 1)) / 4;
		a = (U64_32_t*)((char*)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
		for (;;)
		{
			u.i = a->i;
			u.c[s] = o;
			v.i = u.i;
			v.c[s] = n;
			if (_aso_cas64(&a->i, u.i, v.i) == u.i)
				break;
			for (i = 0;; i++)
				if (i >= elementsof(u.c))
					return a->c[s];
				else if (i != s && u.c[i] != a->c[i])
					break;
		}
		return o;
	}
#endif
	k = lock(state.data, 0, p);
	if (*p == o)
		*p = n;
	else
		o = *p;
	lock(state.data, k, p);
	return o;
}

#ifdef _ast_int8_t

uint64_t
asocas64(uint64_t volatile* p, uint64_t o, uint64_t n)
{
	ssize_t		k;

#if defined(_aso_cas64)
	if (!state.lockf)
		return _aso_cas64(p, o, n);
#endif
	k = lock(state.data, 0, p);
	if (*p == o)
		*p = n;
	else
		o = *p;
	lock(state.data, k, p);
	return o;
}

#endif

/*
 * compare with old, swap with new if same, and return old value
 */

void*
asocasptr(void volatile* p, void* o, void* n)
{
	ssize_t		k;

#if defined(_aso_casptr)
	if (!state.lockf)
		return _aso_casptr((void**)p, o, n);
#endif
	k = lock(state.data, 0, p);
	if (*(void* volatile*)p == o)
		*(void* volatile*)p = n;
	else
		o = *(void* volatile*)p;
	lock(state.data, k, p);
	return o;
}
